nixos/modules/system/etc/etc-activation.nix

```diff
@@ -74,20 +76,64 @@
             ];
           }
         ];
-        services = lib.mkIf config.system.etc.overlay.mutable {
-          rw-etc = {
-            unitConfig = {
-              DefaultDependencies = false;
-              RequiresMountsFor = "/sysroot";
+        services = lib.mkMerge [
+          (lib.mkIf config.system.etc.overlay.mutable {
+            rw-etc = {
+              unitConfig = {
+                DefaultDependencies = false;
+                RequiresMountsFor = "/sysroot";
+              };
+              serviceConfig = {
+                Type = "oneshot";
+                ExecStart = ''
+                  /bin/mkdir -p -m 0755 /sysroot/.rw-etc/upper /sysroot/.rw-etc/work
+                '';
+              };
             };
-            serviceConfig = {
-              Type = "oneshot";
-              ExecStart = ''
-                /bin/mkdir -p -m 0755 /sysroot/.rw-etc/upper /sysroot/.rw-etc/work
+          })
+          {
+            find-etc-metadata-image = {
+              description = "Find the path to the etc metadata image";
+              unitConfig = {
+                DefaultDependencies = false;
+                RequiresMountsFor = "/sysroot/nix/store";
+              };
+              serviceConfig = {
+                Type = "oneshot";
+                RemainAfterExit = true;
+              };
+
+              script = /* bash */ ''
+                set -uo pipefail
+
+                # Figure out what closure to boot
+                closure=
+                for o in $(< /proc/cmdline); do
+                    case $o in
+                        init=*)
+                            IFS='=' read -r -a initParam <<< "$o"
+                            closure="''${initParam[1]}"
+                            ;;
+                    esac
+                done
+
+                # Sanity check
+                if [ -z "''${closure:-}" ]; then
+                  echo 'No init= parameter on the kernel command line' >&2
+                  exit 1
+                fi
+
+                # Resolve symlinks in the init parameter
+                closure="$(chroot /sysroot ${lib.getExe' pkgs.coreutils "realpath"} "$closure")"
+                # Assume the directory containing the init script is the closure.
+                closure="$(dirname "$closure")"
```

Comment:
This is basically the same exact logic as what we have for the activation service, where we find the path to `prepare-root` in the same way.
Maybe it would make sense to create a common service that does this ones and makes the closure available in the initrd so that other units can use it?

Reply:
@ElvishJerricco I'd like to hear your thoughts on this. Would it make sense to pull this into a separate service that creates a link to the closure so that we can reuse this elsewhere?

Reply:
> Would it make sense to pull this into a separate service that creates a link to the closure so that we can reuse this elsewhere?

I think this would be my preferred solution. We could just mount it to `/run/next-system` or something similar.

Reply:
Done, I put this in a separate commit, let me know if you'd prefer a separate PR instead.